File: Sem 1 & 2/Programming/Other_Sem_1/BST/Tree.cpp
#include "Tree.h"
#include <iostream>

using namespace std;

Tree::Tree()
{
	root = nullptr;
}

Tree::Item::Item(int x)
{
	info = x;
	lSon = rSon = father = nullptr;
};

bool Tree::Find(int x, Item *&p)
{
	p = root;
	Item *q = p;
	while (q != nullptr)
	{
		p = q;
		if ((*q).info == x)
		{
			return true;
		}

		if ((*q).info < x)
		{
			q = (*q).rSon;
		}
		else
		{
			q = (*q).lSon;
		}
	}

	return false;
}

bool Tree::Find(int x)
{
	Item *p;
	return Find(x, p);
}

bool Tree::Insert(int x)
{
	Item *r, *p; 

	if (root == nullptr)
	{
		r = new Item(x);
		root = r;
		return true;
	}

	if (Find(x, r))
	{
		return false;
	}

	p = new Item(x);
	(*p).father = r;
	if ((*r).info < x)
	{
		(*r).rSon = p;
	}
	else
	{
		(*r).lSon = p;
	}

	return true;
}

void Tree::deleteItem(Item *x)
{
	if ((*x).father == nullptr)
	{		
		if ((*x).lSon != nullptr) 
		{
			root = (*x).lSon;
			(*(*x).lSon).father = nullptr;
		}
		else 
		{
			root = (*x).rSon;
			if ((*x).rSon != nullptr)
			{
				(*(*x).rSon).father = nullptr;
			}
		}
	}

	else
	{
		if ((*(*x).father).lSon == x)
		{
			if ((*x).lSon != nullptr) 
			{
				(*(*x).father).lSon = (*x).lSon;
				(*(*x).lSon).father = (*x).father;
			}
			else 
			{
				(*(*x).father).lSon = (*x).rSon;
				if ((*x).rSon != nullptr)
				{
					(*(*x).rSon).father = (*x).father;
				}
			}
		}
		else
		{
			if ((*x).lSon != nullptr) 
			{
				(*(*x).father).rSon = (*x).lSon;
				(*(*x).lSon).father = (*x).father;
			}
			else 
			{
				(*(*x).father).rSon = (*x).rSon;
				if ((*x).rSon != nullptr)
				{
					(*(*x).rSon).father = (*x).father;
				}
			}
		}
	}

	(*x).father = (*x).lSon = (*x).rSon = nullptr;

	delete x;
}

bool Tree::Delete(int x)
{
	Item *r, *p;
	if (!Find(x, r))
	{
		return false;
	}

	if (((*r).lSon == nullptr) || ((*r).rSon == nullptr))
	{
		deleteItem(r);
		return true;
	}

	p = (*r).lSon;

	while ((*p).rSon != nullptr)
	{
		p = (*p).rSon;
	}

	(*r).info = (*p).info;
	deleteItem(p);
	return true;
}

Tree::~Tree()
{
	while (root)
	{
		deleteItem(root);
	}

	delete root;
}

File: Sem 1 & 2/Programming/Labs_Sem_2/BST/Source/Tree.cs
using System.Collections.Generic;

namespace BinaryTree
{
    public class Tree
    {
        private Item root;


        public Tree()
        {
            root = null;
        }


        public bool Find(int x)
        {
            Item bufferItem;

            return HiddenFind(x, out bufferItem);
        }

        public bool Insert(int x)
        {
            Item r, p;

            if (root == null)
            {
                r = new Item(x);
                root = r;
                return true;
            }

            if (HiddenFind(x, out r)) return false;

            p = new Item(x);

            p.father = r;

            if (r.data < x)
                r.son_right = p;
            else
                r.son_left = p;

            return true;
        }

        public bool Delete(int x)
        {
            Item r, p;

            if (!HiddenFind(x, out r)) return false;

            if ((r.son_left == null) || (r.son_right == null))
            {
                DeleteItem(r);
                return true;
            }

            p = r.son_left;

            while (p.son_right != null) p = p.son_right;

            r.data = p.data;

            DeleteItem(p);

            return true;
        }


        public List<int> GetLayer(int x)
        {
            List<int> Layer = new List<int>();

            Item
                bufferItem;
            int
                depth = 0,
                current_level = 0;

            if (!HiddenFind(x, out bufferItem)) return Layer;


            while (bufferItem != null)
            {
                bufferItem = bufferItem.father;
                ++depth;
            }

            Queue<Item> ItemQueue = new Queue<Item>();

            ItemQueue.Enqueue(root);
            current_level = 1;

            while (current_level != depth)
            {
                int CurrentLayerCount = ItemQueue.Count;

                for (int i = 1; i <= CurrentLayerCount; i++)
                {
                    var item = ItemQueue.Dequeue();

                    if (item.son_left != null) ItemQueue.Enqueue(item.son_left);
                    if (item.son_right != null) ItemQueue.Enqueue(item.son_right);
                }

                current_level++;
            }

            foreach (var item in ItemQueue)
            {
                Layer.Add(item.data);
            }

            return Layer;
        }



        private void DeleteItem(Item x)
        {
            if (x.father == null)
            {
                if (x.son_left != null)
                {
                    root = x.son_left;
                    x.son_left.father = null;
                }
                else
                {
                    root = x.son_right;
                    if (x.son_right != null)
                        x.son_right.father = null;
                }
            }

            else if (x.father.son_left == x)
            {
                if (x.son_left != null)
                {
                    x.father.son_left = x.son_left;
                    x.son_left.father = x.father;
                }
                else
                {
                    x.father.son_left = x.son_right;
                    if (x.son_right != null)
                        x.son_right.father = x.father;
                }
            }

            else if (x.son_left != null)
            {
                x.father.son_right = x.son_left;
                x.son_left.father = x.father;
            }

            else
            {
                x.father.son_right = x.son_right;
                if (x.son_right != null)
                    x.son_right.father = x.father;
            }

            x.father = x.son_left = x.son_right = null;
        }

        private bool HiddenFind(int x, out Item parameterItem)
        {
            parameterItem = root;

            Item bufferItem = parameterItem;

            while (bufferItem != null)
            {
                parameterItem = bufferItem;

                if (bufferItem.data == x)
                    return true;

                if (bufferItem.data < x)
                    bufferItem = bufferItem.son_right;

                else
                    bufferItem = bufferItem.son_left;
            }

            return false;
        }


        private class Item
        {
            public int data;
            public Item son_left, son_right, father;

            public Item(int x)
            {
                data = x;
                son_left = son_right = father = null;
            }
        }
    }
}
